File: packages/js/core/tests/contract.test.ts
/**
 * EchoButler contract-test runner (TypeScript).
 *
 * Reads the shared `contract-tests/contract-spec.json` and drives the real
 * `@echobutler/core` HTTP binding (`EchoButlerClient.request`) against the
 * docker-compose fixture (`fixture-api` on 127.0.0.1:18080). The mood and
 * stellar high-level wrappers that are already spec-compliant (`@echobutler/mood`,
 * `@echobutler/stellar`) are exercised too.
 *
 * Known wrapper drift (documented in contract-tests/README.md and intentionally
 * NOT asserted as passing):
 *   - `@echobutler/stellar` getTransactionHistory sends `publicKey` as the query
 *     param (canonical wire param is `public_key`).
 *   - `@echobutler/social` LeaderboardClient requests `?window=weekly` and expects
 *     a bare array; canonical is `?limit=` + `{ "entries": [...] }`.
 * The runner exercises those two endpoints at the transport level so the command
 * surface stays covered without masking the drift.
 *
 * The suite self-skips when the fixture is not reachable.
 *
 * Env overrides:
 *   ECHOBUTLER_CONTRACT_SPEC       path to contract-spec.json
 *   ECHOBUTLER_CONTRACT_API_BASE   e.g. http://127.0.0.1:18080
 */
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'
import { EchoButlerClient, EchoButlerError } from '../src'
import type { EchoButlerClient as CoreClient } from '@echobutler/core'
import { getMoodStreak, getMoodSummary, logMood } from '../../mood/src/index'
import { getBalance, submitTransaction } from '../../stellar/src/echobutler'
import { GlobalFeedClient } from '../../social/src/feed'

type Spec = {
  operations: Array<{
    id: string
    method: string
    path: string
    request?: { body?: unknown }
    response: { status: number; body: Record<string, unknown> }
    assertions: Array<{ field: string; eq: unknown; path?: string }>
  }>
}

const defaultSpecPath = fileURLToPath(
  new URL('../../../../contract-tests/contract-spec.json', import.meta.url),
)
const spec = JSON.parse(readFileSync(process.env.ECHOBUTLER_CONTRACT_SPEC ?? defaultSpecPath, 'utf8')) as Spec

const apiBase = process.env.ECHOBUTLER_CONTRACT_API_BASE ?? 'http://127.0.0.1:18080'
const publicKey = 'GDKUJHNOCQ6NOFJCSPE5IZMFFRZ6U4VO3EEFJQKJSDK5B4VZTH4XKSKD'
const client = new EchoButlerClient({
  apiKey: 'contract-test-key',
  baseUrl: apiBase,
  network: 'testnet',
  timeout: 5000,
}) as unknown as CoreClient

function op(id: string) {
  const found = spec.operations.find((o) => o.id === id)
  if (!found) throw new Error(`operation ${id} not found in contract spec`)
  return found
}

/** Walk a dotted path (`entries.0.score`) into a parsed JSON value. */
function at(node: unknown, path: string | undefined): unknown {
  if (!path) return node
  let current: unknown = node
  for (const seg of path.split('.')) {
    if (!seg) continue
    if (Array.isArray(current)) current = current[Number(seg)]
    else current = (current as Record<string, unknown>)[seg]
  }
  return current
}

function assertWire(wire: unknown, assertions: Spec['operations'][number]['assertions'], opId: string) {
  for (const a of assertions) {
    const actual = (at(wire, a.path) as Record<string, unknown>)?.[a.field]
    expect(actual, `${opId}: ${a.path ? a.path + '.' : ''}${a.field}`).toEqual(a.eq)
  }
}

const enabled = await (async () => {
  try {
    const res = await fetch(`${apiBase}/mood/streak`, { signal: AbortSignal.timeout(2000) })
    return res.ok || res.status !== 0
  } catch (err) {
    if (process.env.ECHOBUTLER_CONTRACT_SPEC) {
      throw new Error(
        `contract fixture not reachable at ${apiBase} — contract tests are required because ` +
          `ECHOBUTLER_CONTRACT_SPEC is set: ${(err as Error).message}`,
      )
    }
    return false
  }
})()

describe.skipIf(!enabled)('EchoButler contract (JS binding)', () => {
  it('fetch_mood_streak matches the contract', async () => {
    const streak = await getMoodStreak(client)
    assertWire(streak, op('fetch_mood_streak').assertions, 'fetch_mood_streak')
  })

  it('fetch_mood_summary matches the contract', async () => {
    const summary = await getMoodSummary(client, 'week')
    assertWire(summary, op('fetch_mood_summary').assertions, 'fetch_mood_summary')
  })

  it('log_mood matches the contract', async () => {
    const entry = await logMood(client, { score: 8, note: 'Great day', tags: ['work', 'proud'] })
    assertWire(entry, op('log_mood').assertions, 'log_mood')
  })

  it('get_social_feed matches the contract', async () => {
    const feed = await client.request('GET', '/social/feed?limit=10')
    assertWire(feed, op('get_social_feed').assertions, 'get_social_feed')
  })

  it('get_social_feed_since matches the contract (backfill after a WS gap)', async () => {
    const feedClient = new GlobalFeedClient(client)
    const feed = await feedClient.fetchSince('feed-001', { limit: 50 })
    assertWire(feed, op('get_social_feed_since').assertions, 'get_social_feed_since')
  })

  it('get_leaderboard matches the contract at the transport level', async () => {
    const body = await client.request('GET', '/social/leaderboard?limit=10')
    assertWire(body, op('get_leaderboard').assertions, 'get_leaderboard')
  })

  it('build_echo_transfer matches the contract', async () => {
    const o = op('build_echo_transfer')
    const result = await client.request('POST', o.path, o.request?.body)
    assertWire(result, o.assertions, 'build_echo_transfer')
  })

  it('submit_payment_transaction matches the contract', async () => {
    const o = op('submit_payment_transaction')
    const xdr = (o.request?.body as { xdr: string }).xdr
    const tx = await submitTransaction(client, xdr)
    assertWire(tx, o.assertions, 'submit_payment_transaction')
  })

  it('get_transaction_history matches the contract at the transport level', async () => {
    const body = await client.request(
      'GET',
      `/stellar/transactions?public_key=${publicKey}&limit=10`,
    )
    assertWire(body, op('get_transaction_history').assertions, 'get_transaction_history')
  })

  it('get_stellar_balance_api matches the contract', async () => {
    const balance = await getBalance(client, publicKey)
    assertWire(balance, op('get_stellar_balance_api').assertions, 'get_stellar_balance_api')
  })

  it('api_request_to_unknown_route_must_fail surfaces a 404 error', async () => {
    const o = op('api_request_to_unknown_route_must_fail')
    let caught: unknown
    try {
      await client.request('GET', o.path)
    } catch (err) {
      caught = err
    }
    expect(caught).toBeInstanceOf(EchoButlerError)
    expect((caught as EchoButlerError).statusCode).toBe(404)
  })
})

File: packages/flutter/test/contract_test.dart
// EchoButler contract-test runner (Flutter).
//
// Reads the shared `contract-tests/contract-spec.json` and drives the real
// Flutter bindings (`EchoButler.initialize` + MoodClient/SocialClient/
// StellarClient) against the docker-compose fixture (`fixture-api` on
// 127.0.0.1:18080). Expected values come straight from the spec — this file
// contains no hard-coded numbers, so it can't drift away from the contract.
//
// The suite self-skips when the fixture is not reachable. Build-transfer has
// no Flutter SDK method yet, so it is exercised at the transport level.
//
// Env overrides:
//   ECHOBUTLER_CONTRACT_SPEC       path to contract-spec.json
//   ECHOBUTLER_CONTRACT_API_BASE   e.g. http://127.0.0.1:18080
library;

import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:echobutler_sdk/echobutler_sdk.dart';

Map<String, dynamic> _spec() {
  final path = Platform.environment['ECHOBUTLER_CONTRACT_SPEC'] ??
      '../../contract-tests/contract-spec.json';
  return jsonDecode(File(path).readAsStringSync()) as Map<String, dynamic>;
}

Map<String, dynamic> op(Map<String, dynamic> spec, String id) =>
    (spec['operations'] as List<dynamic>)
        .cast<Map<String, dynamic>>()
        .firstWhere((o) => o['id'] == id);

Future<bool> _fixtureReachable(String base) async {
  try {
    final client = HttpClient();
    client.connectionTimeout = const Duration(seconds: 1);
    final req = await client.getUrl(Uri.parse('$base/mood/streak'));
    final res = await req.close();
    await res.drain<void>();
    client.close();
    return true;
  } catch (_) {
    return false;
  }
}

void main() async {
  final spec = _spec();
  final apiBase = Platform.environment['ECHOBUTLER_CONTRACT_API_BASE'] ??
      'http://127.0.0.1:18080';
  final live = await _fixtureReachable(apiBase);
  if (!live && Platform.environment['ECHOBUTLER_CONTRACT_SPEC'] != null) {
    throw StateError(
      'contract fixture not reachable at $apiBase — contract tests are required because '
      'ECHOBUTLER_CONTRACT_SPEC is set',
    );
  }
  final skipReason = live ? false : 'fixture not reachable at $apiBase';

  final publicKey = ((spec['fixture']['users']['stellar']
          as Map<String, dynamic>)['public_key'])
      .toString();

  await EchoButler.initialize(
    apiKey: (spec['fixture']['api_key'] as String),
    baseUrl: apiBase,
    network: StellarNetwork.testnet,
  );

  group('EchoButler contract (Flutter binding)', () {
    test('fetch_mood_streak matches the contract', () async {
      final streak = await EchoButler.instance.mood.getStreak();
      final body = op(spec, 'fetch_mood_streak')['response']['body']
          as Map<String, dynamic>;
      expect(streak.current, body['current']);
      expect(streak.longest, body['longest']);
      expect(streak.isActiveToday, body['is_active_today']);
      expect(streak.lastLoggedAt,
          DateTime.parse(body['last_logged_at'] as String).toUtc());
    }, skip: skipReason);

    test('fetch_mood_summary matches the contract', () async {
      final summary = await EchoButler.instance.mood.getSummary(period: 'week');
      final body = op(spec, 'fetch_mood_summary')['response']['body']
          as Map<String, dynamic>;
      expect(summary.period, body['period']);
      expect(summary.average, body['average']);
      expect(summary.min, body['min']);
      expect(summary.max, body['max']);
      expect(summary.totalEntries, body['total_entries']);
      expect(summary.trend, body['trend']);
    }, skip: skipReason);

    test('log_mood matches the contract', () async {
      final entry = await EchoButler.instance.mood.log(
        score: 8,
        note: 'Great day',
        tags: const ['work', 'proud'],
      );
      final body =
          op(spec, 'log_mood')['response']['body'] as Map<String, dynamic>;
      expect(entry.id, body['id']);
      expect(entry.userId, body['user_id']);
      expect(entry.score, body['score']);
      expect(entry.note, body['note']);
      expect(entry.tags, body['tags']);
      expect(entry.createdAt,
          DateTime.parse(body['created_at'] as String).toUtc());
    }, skip: skipReason);

    test('get_social_feed matches the contract', () async {
      final feed = await EchoButler.instance.social.getGlobalFeed(limit: 10);
      final entry = op(spec, 'get_social_feed')['response']['body']['entries']
          [0] as Map<String, dynamic>;
      expect(feed.single.id, entry['id']);
      expect(feed.single.score, entry['score']);
      expect(
        feed.single.createdAt,
        DateTime.parse(entry['created_at'] as String).toUtc(),
      );
    }, skip: skipReason);

    test('get_leaderboard matches the contract', () async {
      final leaderboard =
          await EchoButler.instance.social.getLeaderboard(limit: 10);
      final entry = op(spec, 'get_leaderboard')['response']['body']['entries']
          [0] as Map<String, dynamic>;
      expect(leaderboard.single.rank, entry['rank']);
      expect(leaderboard.single.userId, entry['user_id']);
      expect(leaderboard.single.displayName, entry['display_name']);
      expect(leaderboard.single.weeklyScore, entry['weekly_score']);
    }, skip: skipReason);

    test('build_echo_transfer matches the contract at the transport level',
        () async {
      final o = op(spec, 'build_echo_transfer');
      final body = o['request']['body'] as Map<String, dynamic>;
      final expected = o['response']['body'] as Map<String, dynamic>;

      final res = await http.post(
        Uri.parse('$apiBase${o['path']}'),
        headers: {'content-type': 'application/json'},
        body: jsonEncode(body),
      );
      expect(res.statusCode, o['response']['status']);
      final got = jsonDecode(res.body) as Map<String, dynamic>;
      expect(got['xdr'], expected['xdr']);
      expect(got['fee'], expected['fee']);
      expect(got['sequence'], expected['sequence']);
    }, skip: skipReason);

    test('get_stellar_balance_api matches the contract', () async {
      final balance = await EchoButler.instance.stellar.getBalance(publicKey);
      final body = op(spec, 'get_stellar_balance_api')['response']['body']
          as Map<String, dynamic>;
      expect(balance.xlm, body['xlm']);
      expect(balance.echo, body['echo']);
      expect(balance.network, body['network']);
    }, skip: skipReason);

    test('get_transaction_history matches the contract', () async {
      final txs = await EchoButler.instance.stellar
          .getTransactionHistory(publicKey, limit: 10);
      final entry = op(spec, 'get_transaction_history')['response']['body']
          ['transactions'][0] as Map<String, dynamic>;
      expect(txs.single.id, entry['id']);
      expect(txs.single.type, entry['type']);
      expect(txs.single.amount, entry['amount']);
    }, skip: skipReason);

    test('api_request_to_unknown_route_must_fail surfaces a 404 error', () {
      expect(
        EchoButler.instance.mood.getSummary(period: 'contract-unknown-variant'),
        throwsA(isA<EchoButlerError>()),
      );
    }, skip: skipReason);
  });
}
